### PR TITLE
feat(python): support bytecode identification of `numpy` functions in UDFs that we can map to native expressions

### PR DESCRIPTION
**Pull request by alexander-beedie**
Ref: #9968.
Can now identify `numpy` calls in `apply` function bytecode and offer a native polars translation... :)

Understands both direct _"numpy.func(x)"_ calls and the standard _"np.func(x)"_ alias.

Note that I have only whitelisted a handful of numpy functions to start with; it's trivial to add more (literally just add the function name to the whitelist), but each one needs verification that we are indeed doing exactly the same thing in the default case - I'll whitelist additional functions as I confirm this.

## Example

```python
import polars as pl
import numpy as np

df = pl.DataFrame({"abc": [27, 64, 125]})
fn = lambda x: 1 + np.cbrt(x) / 3

df.with_columns(
    val = pl.col("abc").apply(fn),
)

# PolarsInefficientApplyWarning: 
# Expr.apply is significantly slower than the native expressions API.
# Only use if you absolutely CANNOT implement your logic otherwise.
# In this case, you can replace your `apply` with an expression:
#   -  pl.col("abc").apply(lambda x: ...)
#   +  1 + (pl.col("abc").cbrt() / 3)
```
Underlying function bytecode...
```python
import dis
dis.dis( fn )
# 1           0 RESUME                   0
#             2 LOAD_CONST               1 (1)
#             4 LOAD_GLOBAL              0 (np)
#            16 LOAD_METHOD              1 (cbrt)
#            38 LOAD_FAST                0 (x)
#            40 PRECALL                  1
#            44 CALL                     1
#            54 LOAD_CONST               2 (3)
#            56 BINARY_OP               11 (/)
#            60 BINARY_OP                0 (+)
#            64 RETURN_VALUE
```
These instructions get rewritten on initial disassembly with a new synthetic `POLARS_EXPRESSION` which is treated as a special-case unary operator during normal parsing:
```python
# Instruction(opname='LOAD_CONST', opcode=100, arg=1, argval=1, argrepr='1', offset=2, ...
# Instruction(opname='LOAD_FAST', opcode=124, arg=0, argval='x', argrepr='x', offset=26, ...
# Instruction(opname='POLARS_EXPRESSION', opcode=116, arg=1, argval='cbrt', argrepr='cbrt', ...
# Instruction(opname='LOAD_CONST', opcode=100, arg=2, argval=3, argrepr='3', offset=42, ...
# Instruction(opname='BINARY_OP', opcode=122, arg=11, argval=11, argrepr='/', offset=44, ...
# Instruction(opname='BINARY_OP', opcode=122, arg=0, argval=0, argrepr='+', offset=48, ...
```
Limitation: currently only works on single-parameter calls.

### CODE REVIEW

File: py-polars/polars/utils/udfs.py [hunk 1]
```diff
@@ -159,6 +201,8 @@ def _expr(cls, value: StackEntry, col: str, param_name: str, depth: int) -> str:
             op = value.operator
             e1 = cls._expr(value.left_operand, col, param_name, depth + 1)
             if value.operator_arity == 1:
+                if op.isidentifier():
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:204]: this slightly scares me 😄 is there a slightly more robust way?

does this match what's in `NUMPY_MODULE_ALIASES` and `_NUMPY_FUNCTIONS`? If so, could `op` be matched against them directly?

Comment (MarcoGorelli) [reply py-polars/polars/utils/udfs.py:204]: just tried replacing this with `if op in _NUMPY_FUNCTIONS` and the tests pass

File: py-polars/tests/unit/operations/test_inefficient_apply.py [hunk 2]
```diff
@@ -39,14 +40,16 @@ def test_non_simple_function(func: Callable[[Any], Any]) -> None:
         lambda x: not (x > 1) or x == 2,
         lambda x: x is None,
         lambda x: x is not None,
-        lambda x: (x * -x) ** x,
-        lambda x: x * (x**x),
+        lambda x: ((x * -x) ** x) * 1.0,
+        lambda x: 1.0 * (x * (x**x)),
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply py-polars/tests/unit/operations/test_inefficient_apply.py:44]: ah I see, you're using `assert_frame_equal` now which checks the dtypes too - all good then

File: py-polars/polars/utils/udfs.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py:204]: > this slightly scares me 😄 

Lol... it's actually guaranteed to work at this point; however, the _best_ check is probably `op not in _UNARY_OPERATORS`. The only alternative if it fails that check and `operator_arity == 1` is that it _must_ be one of the new synthetic `POLARS_EXPRESSION` ops mapping a function to an expression (I'll add more than numpy matching soon, as well as many more numpy functions, so we don't really want to enumerate all of the explicit matches).

Will tweak accordingly :)

**Review by MarcoGorelli — APPROVED** [review]
Awesome!

BTW, it should be possible to rewrite `pl.col('a').apply(np.sin)` in the same way, right?

**Comment by alexander-beedie** [comment]
> it should be possible to rewrite `pl.col('a').apply(np.sin)` in the same way, right?

We'd need a different mechanism to identify those as they can't be disassembled, but I think it should be pretty easy 👍

**Pull request merged**